### PR TITLE
DPC++ support for oneDAL data management

### PR DESCRIPTION
**Pull request by michael-smirnov**
In this PR:
- DPC++ support for tables, builders and accessors added
- changed API of `homogen_table_builder` to bring more readability (introduced several building methods)
- move ctor of `table_builder` removed (need to add it again later, here I don't have time to implement it)
- `array` class API changes: 
   - `array(int)` ctor replaced by static `empty()` method
   - `unique()` name and behavior changed to `need_mutable_data()`
   - removed behavior with data ownership (methods `is_data_owner()`, `reset_not_owning()`, etc.). Now ownership is ruled only by deleters
   - added aliasing constructor for non-owned data
   - `capacity` property removed, now its always equal to size of an array

:warning: DPC++ implementation for tables and builders need to be tested after we enable DPC++ build support (#686 )

### CODE REVIEW

File: cpp/oneapi/dal/data/detail/access_iface.hpp [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+struct column_values_block {
+    std::int64_t column_index;
+    range rows;
+
+    column_values_block(std::int64_t idx, const range& rows)
+        : column_index(idx),
+          rows(rows) {}
+};
+
+template <typename DataType,
```
[the remark this answers is not in the document]
Comment (rlnx) [reply cpp/oneapi/dal/data/detail/access_iface.hpp:39]: Or if you want to capture meaning use `Data` instead `T`

File: cpp/oneapi/dal/data/detail/access_wrapper_host.hpp [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    }
+
+    template <typename Block>
+    void pull_column(const policy_t&,
+                     Block& block, const column_values_block& index,
+                     const alloc_kind_t&) const {
+        if constexpr (has_pull_column_host<T, typename Block::data_t>::value) {
+            obj_.pull_column(block, index.column_index, index.rows);
+        } else {
+            throw std::runtime_error("pulling column is not supported");
```
Comment (PetrovKP) [inline cpp/oneapi/dal/data/detail/access_wrapper_host.hpp:54]: Will this type of exception be changed after the request https://github.com/oneapi-src/oneDAL/pull/676?

Comment (michael-smirnov) [reply cpp/oneapi/dal/data/detail/access_wrapper_host.hpp:54]: Yes, I especially marked TODO at the head of the file: https://github.com/oneapi-src/oneDAL/blob/b5810686c9a87fd4b33e9b332aece3e6f96fcc87/cpp/oneapi/dal/data/detail/access_wrapper_host.hpp#L22

File: cpp/oneapi/dal/data/detail/access_iface.hpp [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+          typename BlockIndexType,
+          typename PolicyType = detail::host_seq_policy,
+          typename AllocKind = detail::host_only_alloc>
+class access_pull_iface {
+public:
+    virtual ~access_pull_iface() {}
+    virtual void pull(const PolicyType& policy,
+                      array<DataType>& block, const BlockIndexType& index,
```
[the remark this answers is not in the document]
Comment (michael-smirnov) [reply cpp/oneapi/dal/data/detail/access_iface.hpp]: Redesigned - please look again

File: cpp/oneapi/dal/data/detail/access_iface.hpp [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+class access_push_iface {
+public:
+    virtual ~access_push_iface() {}
+    virtual void push(const PolicyType& policy,
+                      const array<DataType>& block, const BlockIndexType& index,
+                      const AllocKind& kind) = 0;
+};
+
+template <typename PolicyType, typename AllocKind>
+struct access_iface {
```
[the remark this answers is not in the document]
Comment (michael-smirnov) [reply cpp/oneapi/dal/data/detail/access_iface.hpp:37]: Redesigned - please look again

File: cpp/oneapi/dal/data/detail/access_iface_type_traits.hpp [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+#pragma once
+
+#include "oneapi/dal/data/array.hpp"
+#include "oneapi/dal/util/detail/type_traits.hpp"
+
+namespace oneapi::dal::detail {
+
+template <typename T, typename DataType>
+struct has_pull_rows_host {
+    INSTANTIATE_HAS_METHOD_DEFAULT_CHECKER(void, pull_rows, (array<DataType>&, const range&) const)
```
[the remark this answers is not in the document]
Comment (michael-smirnov) [reply cpp/oneapi/dal/data/detail/access_iface_type_traits.hpp]: Changed to `ONEAPI_DAL_HAS_METHOD_TRAIT`, more shorten version

**Pull request merged**